pypeapp/lib/anatomy.py

```diff
@@ -692,6 +907,401 @@ def format(self, in_data, only_keys=True):
+        return self._keep_updated
+
+    @staticmethod
+    def project_overrides_path(project_name):
+        project_config_items = [
+            project_anatomy_overrides_dir_path(project_name),
+            Roots.roots_filename
+        ]
+        return os.path.sep.join(project_config_items)
+
+    def _project_overrides_path(self):
+        return Roots.project_overrides_path(self.project_name)
+
+    @property
+    def roots(self):
+        if self.parent is None and self.keep_updated:
+            project_name = os.environ.get("AVALON_PROJECT")
+            if self.project_name != project_name:
+                self._project_name = project_name
+
+        if self.project_name != self.loaded_project:
+            self._roots = None
+
+        if self._roots is None:
+            self._roots = self._discover()
+            self.loaded_project = self.project_name
+            # Backwards compatibility
+            if self._roots is None:
+                self._roots = Roots.default_roots(self)
+        return self._roots
+
+    @staticmethod
+    def default_roots_raw():
+        default_roots_path = os.path.normpath(os.path.join(
+            default_anatomy_dir_path(),
+            Roots.roots_filename
+        ))
+        with open(default_roots_path, "r") as default_roots_file:
+            raw_default_roots = json.load(default_roots_file)
+
+        return raw_default_roots
+
+    @staticmethod
+    def default_roots(parent=None):
+        return Roots._parse_dict(Roots.default_roots_raw())
+
+    def _discover(self):
+        ''' Loads root from json.
+        Default roots are loaded if project is not set.
+
+        :rtype: dictionary
+        '''
+
+        # Return default roots if project is not set
+        if self.project_name is None:
+            return Roots.default_roots(self)
+
+        # Return project specific roots
+        project_roots_path = self._project_overrides_path()
+
+        # If path does not exist we assume it is older project without roots
+        if not os.path.exists(project_roots_path):
+            return None
+
+        with open(project_roots_path, "r") as project_roots_file:
+            raw_project_roots = json.load(project_roots_file)
+
+        return self._parse_dict(raw_project_roots, parent=self)
+
+    @staticmethod
+    def _parse_dict(data, key=None, parent_keys=[], parent=None):
+        is_last = False
+        for value in data.values():
+            if isinstance(value, StringType):
+                is_last = True
+                break
+
+        if is_last:
+            return RootItem(data, key, parent_keys, parent=parent)
+
+        output = {}
+        for key, value in data.items():
+            _parent_keys = list(parent_keys)
+            _parent_keys.append(key)
+            output[key] = Roots._parse_dict(value, key, _parent_keys, parent)
+        return output
+
+    @staticmethod
+    def save_project_overrides(project_name, roots_data=None, override=False):
```

Comment:
This whole file is in a dire need of docstrings. But as far as I can see, it's looking very good. Must be documented though.